src/webnn_native/ops/Binary.cpp

```diff
@@ -19,6 +19,76 @@
 
 namespace webnn_native { namespace op {
 
+    MaybeError Binary::CalculateShape() {
+        auto inputShape1 =
+            mInputs[0]->Shape().empty() ? std::vector<int32_t>{1} : mInputs[0]->Shape();
+        auto inputShape2 =
+            mInputs[1]->Shape().empty() ? std::vector<int32_t>{1} : mInputs[1]->Shape();
+
+        auto l1 = inputShape1.size(), l2 = inputShape2.size();
+        bool shape1IsBigger = l1 >= l2;
+        auto maxShape = shape1IsBigger ? inputShape1 : inputShape2;
+        auto minShape = shape1IsBigger ? inputShape2 : inputShape1;
+        std::vector<int32_t> outputShape;
+        if (mOpType == kMatMul) {
+            if (l1 == 1 && l2 == 1) {
+                if (inputShape1 != inputShape2) {
+                    return DAWN_VALIDATION_ERROR(
+                        "The two 1D inputs of Matmul should have the same shape.");
+                }
+                outputShape = {1};
+            }
+            if (l1 == 2 && l2 == 1) {
+                if (inputShape1[1] != inputShape2[0]) {
+                    return DAWN_VALIDATION_ERROR("The input shapes are incompatible.");
+                }
+                outputShape = {inputShape1[0], 1};
+            }
+            if (l1 == 1 && l2 == 2) {
+                if (inputShape1[0] != inputShape2[0]) {
+                    return DAWN_VALIDATION_ERROR("The input shapes are incompatible.");
+                }
+                outputShape = {1, inputShape2[1]};
+            }
+            if (l1 >= 2 && l2 >= 2) {
+                if (inputShape1[l1 - 1] != inputShape2[l2 - 2]) {
+                    return DAWN_VALIDATION_ERROR("The input shapes are incompatible.");
+                }
+                // broadcasting support
+                for (int32_t i = (int32_t)maxShape.size() - 3, j = (int32_t)minShape.size() - 3;
+                     i >= 0 && j >= 0; --i, --j) {
+                    auto maxDim = maxShape[i], minDim = minShape[j];
+                    if (maxDim != minDim && maxDim != 1 && minDim != 1) {
+                        return DAWN_VALIDATION_ERROR(
+                            "Shapes are not compatible for Matmul, broadcasting failed.");
+                    }
+                    if (maxDim < minDim) {
+                        maxShape[i] = minDim;
+                    }
+                }
+                outputShape = maxShape;
+                outputShape[outputShape.size() - 1] = inputShape2[l2 - 1];
+                outputShape[outputShape.size() - 2] = inputShape1[l1 - 2];
+            }
+        } else {
+            // broadcasting support
+            for (int32_t i = (int32_t)maxShape.size() - 1, j = (int32_t)minShape.size() - 1;
+                 i >= 0 && j >= 0; --i, --j) {
+                auto maxDim = maxShape[i], minDim = minShape[j];
```

Comment:
Please follow https://numpy.org/doc/stable/user/basics.broadcasting.html#general-broadcasting-rules that is referred in spec, it says
> It starts with the trailing (i.e. rightmost) dimensions and works its way left. Two dimensions are compatible when

> 1. they are equal, or

> 2. one of them is 1

> If these conditions are not met, a ValueError: operands could not be broadcast together exception is thrown, indicating that the arrays have incompatible shapes. The size of the resulting array is the size that is not 1 along each axis of the inputs.

And you can probably iterate the shapes like this:
```c++
for (size_t i = 0; i < rankOutput; ++i) {
  auto dimA = i < rankA - 1 ? shapeA[rankA - i - 1] : 1;
  auto dimB = i < rankB - 1 ? shapeB[rankB - i - 1] : 1;
  ...
}
```